File: Echo/DownloadLessonViewController.m
//
//  DownloadLessonsViewController.m
//  Echo
//
//  Created by Will Entriken on 2/11/13.
//
//

#import "DownloadLessonViewController.h"
#import "NetworkManager.h"
#import "LanguageSelectController.h"
#import "UIImageView+AFNetworking.h"
#import <MessageUI/MessageUI.h>
#import "GAI.h"
#import "GAIFields.h"
#import "GAIDictionaryBuilder.h"
#import "Echo-Swift.h"

@interface DownloadLessonViewController () <LanguageSelectControllerDelegate, MFMailComposeViewControllerDelegate>
@property (strong, nonatomic) NSArray *lessons;
- (void)populateRowsWithSearch:(NSString *)searchString languageTag:(NSString *)tag;
@end

@implementation DownloadLessonViewController
@synthesize delegate = _delegate;

- (void)populateRowsWithSearch:(NSString *)searchString languageTag:(NSString *)tag
{
    NetworkManager *networkManager = [NetworkManager sharedNetworkManager];
    [networkManager searchLessonsWithLangTag:tag andSearhText:searchString onSuccess:^(NSArray *lessonPreviews)
     {
         self.lessons = lessonPreviews;
         [self.tableView reloadData];
     } onFailure:^(NSError *error) {
         [NetworkManager hudFlashError:error];
     }];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    Profile *me = [Profile currentUserProfile];
    self.navigationItem.rightBarButtonItem.title = me.learningLanguageTag;
    [self populateRowsWithSearch:@"" languageTag:me.learningLanguageTag];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    id tracker = [GAI sharedInstance].defaultTracker;
    [tracker set:kGAIScreenName value:@"DownloadLesson"];
    [tracker send:[[GAIDictionaryBuilder createAppView] build]];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.lessons.count + 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == self.lessons.count)
        return [tableView dequeueReusableCellWithIdentifier:@"request" forIndexPath:indexPath];
    
    Lesson *lesson = (self.lessons)[indexPath.row];
    NetworkManager *networkManager = [NetworkManager sharedNetworkManager];
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"lesson" forIndexPath:indexPath];
    ((UILabel *)[cell viewWithTag:1]).text = lesson.name;
    ((UILabel *)[cell viewWithTag:2]).text = lesson.detail;
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.dateStyle = NSDateFormatterMediumStyle;
    dateFormatter.timeStyle = NSDateFormatterNoStyle;
    NSDate *date = [NSDate dateWithTimeIntervalSince1970:lesson.serverTimeOfLastCompletedSync];
    NSString *formattedDateString = [dateFormatter stringFromDate:date];
    ((UILabel *)[cell viewWithTag:3]).text = formattedDateString;
    if (lesson.numLikes) {
        ((UILabel *)[cell viewWithTag:4]).text = [NSString stringWithFormat:@"%ld", (long)lesson.numLikes];
        [(UIImageView *)[cell viewWithTag:7] setHidden:NO];
    } else {
        ((UILabel *)[cell viewWithTag:4]).text = @"";
        [(UIImageView *)[cell viewWithTag:7] setHidden:YES];
    }
    if (lesson.numFlags) {
        ((UILabel *)[cell viewWithTag:5]).text = [NSString stringWithFormat:@"%ld", (long)lesson.numFlags];
        [(UIImageView *)[cell viewWithTag:8] setHidden:NO];
    } else {
        ((UILabel *)[cell viewWithTag:5]).text = @"";
        [(UIImageView *)[cell viewWithTag:8] setHidden:YES];
    }
    ((UILabel *)[cell viewWithTag:6]).text = lesson.userName;
    UIImage *placeholder = [UIImage imageNamed:@"none40"];
    NSURL *userPhoto = [networkManager photoURLForUserWithID:@(lesson.userID)];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:userPhoto];
    [(UIImageView *)[cell viewWithTag:9] setImageWithURLRequest:request placeholderImage:placeholder success:nil failure:nil];
    return cell;
}

#pragma mark - Table view delegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row < self.lessons.count)
        return 69;
    else
        return tableView.rowHeight;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row < self.lessons.count) {
        [self.delegate downloadLessonViewController:self gotStubLesson:(self.lessons)[indexPath.row]];
        id<GAITracker> tracker = [GAI sharedInstance].defaultTracker;
        [tracker send:[[GAIDictionaryBuilder createEventWithCategory:@"Usage"
                                                              action:@"Learning"
                                                               label:@"Download Lesson"
                                                               value:@(1)] build]];
        
    } else {
        if ([MFMailComposeViewController canSendMail]) {
            MFMailComposeViewController *picker = [[MFMailComposeViewController alloc] init];
            picker.mailComposeDelegate = self;
            [picker setSubject:[NSString stringWithFormat:@"Idea for Echo lesson/%@", self.navigationItem.rightBarButtonItem.title]];
            [picker setToRecipients:@[@"echo@phor.net"]];
            [picker setMessageBody:@"(TYPE YOUR IDEA IN HERE)" isHTML:NO];
            [self presentViewController:picker animated:YES completion:nil];
        }
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    LanguageSelectController *controller = segue.destinationViewController;
    controller.delegate = self;
}

#pragma mark - LanguageSelectControllerDelegate

- (void)languageSelectController:(id)controller didSelectLanguage:(NSString *)tag withNativeName:(NSString *)name
{
    [self populateRowsWithSearch:@"" languageTag:tag];
    self.navigationItem.rightBarButtonItem.title = tag;
    [self.navigationController popToViewController:self animated:YES];
}

#pragma mark - MFMailComposeViewControllerDelegate

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Echo/DownloadLessonViewController.swift
//
//  DownloadLessonViewController.swift
//  Echo
//
//  Created by William Entriken on 1/11/16.
//
//

import Foundation
import MessageUI
import UIKit
import Firebase
import MBProgressHUD
import AlamofireImage

protocol DownloadLessonDelegate: class {
    func downloadLessonViewController(controller: UIViewController, gotStubLesson lesson: Lesson)
}

class DownloadLessonViewController: UITableViewController {
    weak var delegate: DownloadLessonDelegate?
    var lessons = [Lesson]()

    func populateRowsWithSearch(searchString: String, languageTag tag: String) {
        let networkManager: NetworkManager = NetworkManager.sharedNetworkManager
        networkManager.searchLessonsWithLangTag(tag, andSearhText: searchString, onSuccess: { (lessonPreviews: [Lesson]) -> Void in
            self.lessons = lessonPreviews
            self.tableView.reloadData()
            }) {(error: NSError!) -> Void in
                MBProgressHUD.flashError(error)
        }
    }
}

extension DownloadLessonViewController /*: UIViewController */ {
    override func viewDidLoad() {
        super.viewDidLoad()
        let me: Profile = Profile.currentUser
        self.navigationItem.rightBarButtonItem!.title = me.learningLanguageTag
        self.populateRowsWithSearch("", languageTag: me.learningLanguageTag)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        FIRAnalytics.logEventWithName("page_view", parameters: ["name": NSStringFromClass(self.dynamicType)])
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        guard let destination = segue.destinationViewController as? LanguageSelectController else {
            return
        }
        destination.delegate = self
    }
    
}

extension DownloadLessonViewController /*: UITableViewDataSource */ {
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if indexPath.row < self.lessons.count {
            return 69
        }
        else {
            return tableView.rowHeight
        }
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.lessons.count + 1
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        if indexPath.row == self.lessons.count {
            return tableView.dequeueReusableCellWithIdentifier("request", forIndexPath: indexPath)
        }
        let lesson: Lesson = (self.lessons)[indexPath.row]
        let networkManager: NetworkManager = NetworkManager.sharedNetworkManager
        let cell: UITableViewCell = tableView.dequeueReusableCellWithIdentifier("lesson", forIndexPath: indexPath)
        (cell.viewWithTag(1) as! UILabel).text = lesson.name
        (cell.viewWithTag(2) as! UILabel).text = lesson.detail
        let dateFormatter: NSDateFormatter = NSDateFormatter()
        dateFormatter.dateStyle = .MediumStyle
        dateFormatter.timeStyle = .NoStyle
        let date = NSDate(timeIntervalSince1970: Double(lesson.serverTimeOfLastCompletedSync))
        let formattedDateString: String = dateFormatter.stringFromDate(date)
        (cell.viewWithTag(3) as! UILabel).text = formattedDateString
        if lesson.numLikes > 0 {
            (cell.viewWithTag(4) as! UILabel).text = "\(Int(lesson.numLikes))"
            (cell.viewWithTag(7) as! UIImageView).hidden = false
        } else {
            (cell.viewWithTag(4) as! UILabel).text = ""
            (cell.viewWithTag(7) as! UIImageView).hidden = true
        }
        if lesson.numFlags > 0 {
            (cell.viewWithTag(5) as! UILabel).text = "\(Int(lesson.numFlags))"
            (cell.viewWithTag(8) as! UIImageView).hidden = false
        } else {
            (cell.viewWithTag(5) as! UILabel).text = ""
            (cell.viewWithTag(8) as! UIImageView).hidden = true
        }
        (cell.viewWithTag(6) as! UILabel).text = lesson.userName
        let placeholderImage = UIImage(named: "user")!
        let userPhotoUrl = networkManager.photoURLForUserWithID(lesson.userID)
        (cell.viewWithTag(9) as! UIImageView).af_setImageWithURL(
            userPhotoUrl,
            placeholderImage: placeholderImage,
            filter: nil,
            imageTransition: .CrossDissolve(0.2)
        )
        return cell
    }
}

extension DownloadLessonViewController /*: UITableViewDelegate*/ {
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        if indexPath.row < self.lessons.count {
            self.delegate!.downloadLessonViewController(self, gotStubLesson: (self.lessons)[indexPath.row])
            FIRAnalytics.logEventWithName("learning", parameters: ["action": "Download Lesson"])
        }
        else {
            if MFMailComposeViewController.canSendMail() {
                let picker = MFMailComposeViewController()
                picker.mailComposeDelegate = self
                picker.setSubject("Idea for Echo lesson/\(self.navigationItem.rightBarButtonItem!.title)")
                picker.setToRecipients(["echo@phor.net"])
                picker.setMessageBody("(TYPE YOUR IDEA IN HERE)", isHTML: false)
                self.presentViewController(picker, animated: true, completion: nil)
            }
        }
    }
   
}

extension DownloadLessonViewController: LanguageSelectControllerDelegate {
    func languageSelectController(controller: AnyObject, didSelectLanguage tag: String, withNativeName name: String) {
        self.populateRowsWithSearch("", languageTag: tag)
        self.navigationItem.rightBarButtonItem!.title = tag
        self.navigationController!.popToViewController(self, animated: true)
    }
}

extension DownloadLessonViewController: MFMailComposeViewControllerDelegate {
    func mailComposeController(controller: MFMailComposeViewController, didFinishWithResult result: MFMailComposeResult, error: NSError?) {
        self.dismissViewControllerAnimated(true, completion: { _ in })
    }
}